packages/backend-defaults/src/entrypoints/actions/actionsServiceFactory.test.ts

```diff
@@ -0,0 +1,334 @@
+    const mockActionsListEndpoint = jest.fn();
+    const mockNotFoundActionsListEndpoint = jest.fn();
+
+    const defaultServices = [
+      mockServices.rootConfig.factory({
+        data: {
+          backend: {
+            actions: {
+              pluginSources: ['my-plugin', 'not-found-plugin'],
+            },
+          },
+        },
+      }),
+      actionsServiceFactory,
+      httpRouterServiceFactory,
+      mockServices.httpAuth.factory({
+        defaultCredentials: mockCredentials.service('user:default/mock'),
+      }),
+      mockServices.discovery.factory(),
+      actionsRegistryServiceFactory,
+    ];
+
+    const mockActionsDefinition: ActionsServiceAction = {
+      description: 'my mock description',
+      id: 'my-plugin:test',
+      name: 'testy',
+      title: 'Test',
+      schema: {
+        input: {},
+        output: {},
+      },
+    };
+
+    beforeEach(() => {
+      server.use(
+        rest.get(
+          'http://localhost:0/api/my-plugin/.backstage/actions/v1/actions',
+          mockActionsListEndpoint.mockImplementation((_req, res, ctx) =>
+            res(
+              ctx.json({
+                actions: [mockActionsDefinition],
+              }),
+            ),
+          ),
+        ),
+        rest.get(
+          'http://localhost:0/api/not-found-plugin/.backstage/actions/v1/actions',
+          mockNotFoundActionsListEndpoint.mockImplementation((_req, res, ctx) =>
+            res(ctx.status(404)),
+          ),
+        ),
+      );
+    });
+
+    describe('list', () => {
+      it('should list all plugins in config to find actions and handle failures gracefully', async () => {
+        const subject = await ServiceFactoryTester.from(actionsServiceFactory, {
+          dependencies: defaultServices,
+        }).getSubject();
+
+        const { actions } = await subject.list({
+          credentials: mockCredentials.service('user:default/mock'),
+        });
+
+        expect(actions).toEqual([mockActionsDefinition]);
+
+        expect(mockActionsListEndpoint).toHaveBeenCalledTimes(1);
+        expect(mockNotFoundActionsListEndpoint).toHaveBeenCalledTimes(1);
+      });
+    });
+
+    describe('invoke', () => {
+      it('should invoke the action and return the output', async () => {
+        server.use(
+          rest.post(
+            'http://localhost:0/api/my-plugin/.backstage/actions/v1/actions/my-plugin:test/invoke',
+            (_req, res, ctx) => res(ctx.json({ output: { ok: true } })),
+          ),
+        );
+        const subject = await ServiceFactoryTester.from(actionsServiceFactory, {
+          dependencies: defaultServices,
+        }).getSubject();
+
+        const { output } = await subject.invoke({
+          id: 'my-plugin:test',
+          credentials: mockCredentials.service('user:default/mock'),
+        });
+
+        expect(output).toEqual({ ok: true });
+      });
+
+      it('should throw a 404 if the action does not exist', async () => {
```

Comment:
Use `it.each` instead of copying and pasting the test.

```js
it.each([
  [404, 'does not exist'], 
  [400, 'returns an invalid input'],
])('should throw a %n if the action %s',  async (statusCode) => {
  server.use(
    rest.post(
      'http://localhost:0/api/my-plugin/.backstage/actions/v1/actions/my-plugin:test/invoke',
      (_req, res, ctx) => res(ctx.status(statusCode)),
    ),
  );

  const subject = await ServiceFactoryTester.from(actionsServiceFactory, {
    dependencies: defaultServices,
  }).getSubject();

  await expect(
    subject.invoke({
      id: 'my-plugin:test',
      credentials: mockCredentials.service('user:default/mock'),
    }),
  ).rejects.toThrow(String(statusCode));
});
````